File: CodeBustersV1.java
import java.util.*;
import java.io.*;
import java.math.*;

/**
 * Send your busters out into the fog to trap ghosts and bring them home!
 **/
class Player {

    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        int bustersPerPlayer = in.nextInt(); // the amount of busters you control
        int ghostCount = in.nextInt(); // the amount of ghosts on the map
        int myTeamId = in.nextInt(); // if this is 0, your base is on the top left of the map, if it is one, on the bottom right
        String ordres=new String();
        String ordres2=new String();
        ordres="    ";
        ordres2="    ";
        boolean bust=false;
        boolean release=false;
        int myBusterX=0;
        int myBusterY=0;
        int ghostX=0;
        int ghostY=0;
        int myBaseX=0;
        int myBaseY=0;
        if (myTeamId==0) { 
            myBaseX=600;
            myBaseY=600;
            }
        else if (myTeamId==1) {
            myBaseX=15400;
            myBaseY=8400;
        }
        // game loop
        while (true) {
            int entities = in.nextInt(); // the number of busters and ghosts visible to you
            for (int i = 0; i < entities; i++) {
                int entityId = in.nextInt(); // buster id or ghost id
                int x = in.nextInt();
                int y = in.nextInt(); // position of this buster / ghost
                int entityType = in.nextInt(); // the team id if it is a buster, -1 if it is a ghost.
                int state = in.nextInt(); // For busters: 0=idle, 1=carrying a ghost.
                int value = in.nextInt(); // For busters: Ghost id being carried. For ghosts: number of busters attempting to trap this ghost.
                
                ////////////////////////////////////////////////////////////////////DEBUT DE MON CODE////////////////////////////////////////////////////////////////
                
                
                if (entityType==-1) {                                                       // SI C'EST UN F
                    ghostX=x;
                    ghostY=y;
                    int ghostId=entityId;
                    if (((x>900) && (x<1760)) || ((y>900) && (y<1760)) && (bust==false)) {      // S'il est à portée, le bust
                        ordres="BUST "+entityId;
                    }
                    else if (bust==false) {                                                     // Sinon le traquer
                        ordres="MOVE "+(ghostX+901)+" "+(ghostY+901);
                    }
                    else if ((ghostX>2200) && (ghostY>2200)) {                                  // Si bust raté revenir au milieu
                        ordres="MOVE "+8000+" "+4500;
                    }
                }
                
                if (entityType==myTeamId) {                                                 // SI  C'EST UN B ALLIE
                    myBusterX=x;
                    myBusterY=y;
                    if (bust==false) {                                                          // s'il ne fait rien alors qu'il bouge
                        ordres="MOVE "+8000+" "+4500;
                    }
                    if ((bust==true) && (state==1)) {                                           // S'il porte un F, rentrer à la base
                        ordres="MOVE "+myBaseX+" "+myBaseY;
                    }
                    if ((bust==true) && ((x==myBaseX) && (y==myBaseY))) {                               // Le Release à la base
                        ordres="RELEASE";
                    }
                    if (release==true) {
                        if ((myBusterX<15000) && (myBusterY<8000) && (myTeamId==0)) {
                        ordres="MOVE "+(x+400)+" "+(y+300);                                     // après release repartir à la chasse
                        }
                        else if ((myBusterX>=15000) && (myBusterY>=8000) && (myTeamId==0)) {
                        ordres="MOVE "+(x-400)+" "+(y-300);                                     // après release repartir à la chasse
                        }
                    }
                    if ((myBusterX>=16000) && (myBusterY>=9000)) {                              // s'il arrive en coin de map revenir au milieu
                        ordres="MOVE "+15300+" "+8300;
                    }
                    if ((myBusterX==901) && (myBusterY==901)) {
                        ordres="MOVE "+8000+" "+4500;
                    }
                    if ((myBusterX==8000) && (myBusterY==4500)) {
                        ordres="MOVE "+myBaseX+" "+myBaseY;
                    }
                }
                
                if (((entityType!=-1) && (entityType!=myTeamId)) && (state==1)) {           // SI C'EST UN B ENNEMI AVEC UN F
                    if (((x>900) && (x<1760)) || ((y>900) && (y<1760))) {                        // S'il est à portée, le bust
                        ordres="BUST "+entityId;
                    }
                    else if (bust==false) {                                                      // Sinon le traquer
                        ordres="MOVE "+(x+901)+" "+(y+901);
                    }
                }
            }
            for (int i = 0; i < bustersPerPlayer; i++) {
                if (ordres.substring(0,4).equals("BUST")) {
                    bust=true;
                    release=false;
                }
                else if (ordres.equals("RELEASE")) {
                    bust=false;
                    release=true;
                }
                
                System.out.println(ordres); // MOVE x y | BUST id | RELEASE
                
            }
        }
    }
}

File: codebustersV1.js
// CALCULATE DISTANCE BETWEEN 2 ENTITIES
function calcDistance(x1,x2,y1,y2) {
    var distance=Math.sqrt(Math.pow(x2-x1,2)+Math.pow(y2-y1,2));
    return Math.round(distance);
}


var bustersPerPlayer = parseInt(readline()); // the amount of busters you control
var ghostCount = parseInt(readline()); // the amount of ghosts on the map
var myTeamId = parseInt(readline()); // if this is 0, your base is on the top left of the map, if it is one, on the bottom right
if (myTeamId===0) { var myBase={x:500,y:500}; var enemyBase={x:16000,y:8500}; } 
             else { var myBase={x:15500,y:8500}; var enemyBase={x:0,y:500}; }
var myBusters=[];
var enemyBusters=[];
var ghosts=[];
var actions=[];
var distances=[];
var tour=0;

// game loop
while (true) {    var entities = parseInt(readline()); // the number of busters and ghosts visible to you
    
    for (var i = 0; i < entities; i++) {
        var inputs = readline().split(' ');
        var entityId = parseInt(inputs[0]); // buster id or ghost id
        var x = parseInt(inputs[1]);
        var y = parseInt(inputs[2]); // position of this buster / ghost
        var entityType = parseInt(inputs[3]); // the team id if it is a buster, -1 if it is a ghost.
        var state = parseInt(inputs[4]); // For busters: 0=idle, 1=carrying a ghost.
        var value = parseInt(inputs[5]); // For busters: Ghost id being carried. For ghosts: number of busters attempting to trap this ghost.
        // MES BUSTERS
        if (entityType===myTeamId) {
            myBusters.push({id:entityId,posX:x,posY:y,status:state});
        }
        // LES FANTOMES
        if (entityType==-1) {
            ghosts.push({id:entityId,posX:x,posY:y});
            
        }
        // LES BUSTERS ENNEMIS
        if ((entityType!=myTeamId)&&(entityType!==-1)) {
            enemyBusters.push({id:entityId,posX:x,posY:y,status:state});
        }
    }
    
    //////////////////////////////////// DEBUGGING LINES //////////////////////////////////
    printErr("FANTOMES RESTANTS :"+ghostCount+"\n J'AI "+bustersPerPlayer+" BUSTERS. LEURS COORDONNEES SONT:");
    for (var i=0;i<bustersPerPlayer;i++) {
        printErr("BUSTER "+i+"(ID:"+myBusters[i].id+"):"+myBusters[i].posX+" "+myBusters[i].posY+" ETAT:"+myBusters[i].status);
        
    }
    ///////////////////////////////////////////////////////////////////////////////////////
   
    //////////////////////////// ACTION DECISION FOR EACH BUSTER /////////////////////////
    for (var i = 0;i<bustersPerPlayer;i++) {
        // SI LE BUSTER N'A PAS DE PRISONNIER
        if (myBusters[i].status===0) {
            // s'il voit au moins un fantome autour de lui...
            if (ghosts.length>0) {
                var distance=calcDistance(myBusters[i].posX,ghosts[0].posX,myBusters[i].posY,ghosts[0].posY);
                // Si le fantome est dans le range alors BUST...
                if ((distance>900)&&(distance<1760)) {
                    actions.push("BUST "+ghosts[0].id);
                    ghosts.shift();
                }
                // sinon s'il est trop près ou trop loin, MOVE près de lui
                else if ((distance<900)||(distance>1760)) {
                    actions.push("MOVE "+ghosts[0].posX+" "+ghosts[0].posY);
                }
            }
            // sinon si pas de fantome visible alors explorer
            else if (ghosts.length===0) {
                // si ma base est à gauche...
                if (myTeamId===0) {
                   // tant qu'on a pas atteint le bord droit...
                    if (myBusters[i].posX!=enemyBase.x) {
                        // aller vers la droite toute
                        actions.push("MOVE "+(myBusters[i].posX+800)+" "+(myBusters[i].posY+(i*200)));
                    }
                    // une fois le bord droit atteint...
                    else if (myBusters[i].posX==enemyBase.x) {
                        // aller base ennemie
                        actions.push("MOVE "+enemyBase.x+" "+enemyBase.y);
                    } 
                }
                // sinon si ma base est à droite...
                else if (myTeamId===1) {
                   // tant qu'on a pas atteint le bord gauche...
                    if (myBusters[i].posX!=enemyBase.x) {
                        // aller vers la gauche toute
                        actions.push("MOVE "+(myBusters[i].posX-800)+" "+(myBusters[i].posY+(i*100)));
                    }
                    // une fois le bord gauche atteint...
                    else if (myBusters[i].posX==enemyBase.x) {
                        // aller base ennemie
                        actions.push("MOVE "+enemyBase.x+" "+enemyBase.y);
                    } 
                }
            }
        }
        // SI LE BUSTER TRANSPORTE UN FANTOME
        else if (myBusters[i].status===1) {
            // S'il est aux coordonnées de ma base, alors RELEASE...
            if ((myBusters[i].posX==myBase.x)&&(myBusters[i].posY==myBase.y)) {
                actions.push("RELEASE");
            }
            // Sinon, MOVE vers ma base
            else if ((myBusters[i].posX!=myBase.x)||(myBusters[i].posY!=myBase.y)) {
                actions.push("MOVE "+myBase.x+" "+myBase.y);
            } 
        }
    }
    //////////////////////////////////////////////////////////////////////////////////
    
    ////////////////////////// PRINTING TOUR INSTRUCTIONS ////////////////////////////
    for (var x=0,l=actions.length;x<l;x++) {
        print(actions.shift());
    }
    /////////////////////////////////////////////////////////////////////////////////
    
    // EMPTYING ARRAYS BEFORE WHILE LOOP REFRESHING
    myBusters.length=0;
    enemyBusters.length=0;
    ghosts.length=0;
    tour+=2;
    
} // FIN BOUCLE JEU